### PR TITLE
Parent Decoration

### PR DESCRIPTION
**Pull request by bquantump**
https://dev.azure.com/azure-mgmt-ex/DotNET%20Management%20SDK/_boards/board/t/DotNET%20Management%20SDK%20Team/Stories/?workitem=5504

### CODE REVIEW

**Comment by chamons** [comment]
I suspect your branch needs rebased.

Consider recreating the branch and using cherry pick, or git pull --rebase.

**Review by m-nash — CHANGES_REQUESTED** [review]
need to merge in latest feature/v3

**Review by chamons — CHANGES_REQUESTED** [review]
Ok, 93 files and  +333 −76  is a bit more reasonable at least.

In PRs please post a description in the PR of the change, why it was added, anything in particular we should be looking out for etc.

Just posting a issue link is only really fine for trivial "fix bug X" few line fixes.

**Comment by chamons** [comment]
Actual changed non-gen'ed files:

```
src/AutoRest.CSharp/AutoRest/Communication/StandaloneGeneratorRunner.cs
src/AutoRest.CSharp/AutoRest/Plugins/Configuration.cs
src/AutoRest.CSharp/Generation/Writers/DataPlaneLongRunningOperationWriter.cs
src/AutoRest.CSharp/Generation/Writers/DataPlaneResponseHeaderGroupWriter.cs
src/AutoRest.CSharp/Input/CodeModelPartials.cs
src/AutoRest.CSharp/Output/Models/Requests/DataPlaneLongRunningOperationMethod.cs
src/AutoRest.CSharp/Output/Models/Requests/ProviderSegment.cs
src/AutoRest.CSharp/Output/Models/Types/ExtensionDetection.cs
src/AutoRest.CSharp/Output/Models/Types/MgmtOutputLibrary.cs
src/AutoRest.CSharp/Output/Models/Types/ParentDetection.cs
src/AutoRest.CSharp/Output/Models/Types/ProviderSegmentDetection.cs
src/AutoRest.CSharp/Output/Models/Types/TenantDetection.cs
test/AutoRest.TestServer.Tests/TestMgmtParent.cs
test/TestProjects/OperationGroupMappings/readme.md
test/TestProjects/TenantOnly/readme.md
samples/Azure.Management.Storage/readme.md
samples/Azure.ResourceManager.Sample/readme.md
```

**Review by m-nash — APPROVED** [review]

File: src/AutoRest.CSharp/Output/Models/Requests/ProviderSegment.cs
Comment (pakrym) [inline src/AutoRest.CSharp/Output/Models/Requests/ProviderSegment.cs:15]: nit: formatting + accesibility consistency with other properties

File: test/AutoRest.TestServer.Tests/TestMgmtParent.cs
Comment (pakrym) [inline test/AutoRest.TestServer.Tests/TestMgmtParent.cs:32]: @m-nash @allenjzhang this is what I was talking about. To propagate values you now need to do strange things that don't seem related.

File: test/AutoRest.TestServer.Tests/TestMgmtParent.cs
Comment (pakrym) [inline test/AutoRest.TestServer.Tests/TestMgmtParent.cs:24]: Can we directly create models required for these tests instead of loading a sample? Samples are based on real swagger that is not stable and can change in the future. At a minimum, we should use a hand-crafted test project swagger for this.

File: src/AutoRest.CSharp/Output/Models/Types/ParentDetection.cs
Comment (pakrym) [inline src/AutoRest.CSharp/Output/Models/Types/ParentDetection.cs:21]: I haven't seen any coverage for all these exception scenarios.

File: src/AutoRest.CSharp/Output/Models/Types/ExtensionDetection.cs
Comment (pakrym) [inline src/AutoRest.CSharp/Output/Models/Types/ExtensionDetection.cs:12]: Having something that is an extension for the OperationGroup and is later written into a property on the  OperationGroup is double confusing.

File: test/AutoRest.TestServer.Tests/TestMgmtParent.cs
Comment (m-nash) [reply test/AutoRest.TestServer.Tests/TestMgmtParent.cs:32]: ?? not sure I follow

Comment (pakrym) [reply test/AutoRest.TestServer.Tests/TestMgmtParent.cs:32]: Same thing about mutable properties on OperationGroup being assigned as a side effect of accessing the `context.Library`. And dependent on the order of execution.

It also makes it very hard to test this new feature in isolation without loading the entire code model and building a library.

Comment (m-nash) [reply test/AutoRest.TestServer.Tests/TestMgmtParent.cs:32]: this is a test case I don't think I follow

Comment (pakrym) [reply test/AutoRest.TestServer.Tests/TestMgmtParent.cs:32]: Do you mean don't follow the test case of the problem I'm trying to describe?

Comment (m-nash) [reply test/AutoRest.TestServer.Tests/TestMgmtParent.cs:32]: the problem being described

Comment (pakrym) [reply test/AutoRest.TestServer.Tests/TestMgmtParent.cs:32]: 2 problems:
1. this test has to run half of the generator to test a small piece of code because of the way things are coupled
2. The value of the `operations.Parent` property depends on the order of execution and side effects of other parts of the system.

For example:

``` C#
     var model = await codeModelTask;
            var context = new BuildContext<MgmtOutputLibrary>(model, configuration, sourceInputModel);
            _ = context.Library; // gen lib

model.OperationGroups.First().Parent
```

Would give you the correct value. Note that context and library were not used in any of the asserts so let's remove them and try the following code instead:

while 

``` C#
    var model = await codeModelTask;
    
	model.OperationGroups.First().Parent

```

Whoops, it doesn't work anymore.

Comment (m-nash) [reply test/AutoRest.TestServer.Tests/TestMgmtParent.cs:32]: Ok, so basically you are re-asserting that this [item ](https://dev.azure.com/azure-mgmt-ex/DotNET%20Management%20SDK/_workitems/edit/5724)we opened up needs to get fixed.  I agree.

Comment (pakrym) [reply test/AutoRest.TestServer.Tests/TestMgmtParent.cs:32]: No, that one talks about models returned from the output library. THis is about the mutability of the CodeModel types that serve as an input to the OutputLibrary.

**Review by chamons — CHANGES_REQUESTED** [review]
See https://github.com/Azure/autorest.csharp/pull/1117#discussion_r605981621 for a writeup, but I have architectural concerns that I'd like resolved before landing this.

File: src/AutoRest.CSharp/Output/Models/Types/ParentDetection.cs
Comment (m-nash) [reply src/AutoRest.CSharp/Output/Models/Types/ParentDetection.cs:21]: We need to Spoof the creation of an OperationGroup object which I will create an ADO item.
https://dev.azure.com/azure-mgmt-ex/DotNET%20Management%20SDK/_workitems/edit/5755

File: src/AutoRest.CSharp/Output/Models/Types/ExtensionDetection.cs
Comment (m-nash) [reply src/AutoRest.CSharp/Output/Models/Types/ExtensionDetection.cs:12]: Not sure I follow, this isn't an extension method on an operationgroup it is a method which determines whether an operationgroup is an extension resource type in ARM and then sets the result of this as a property on the operationgroup

**Comment by m-nash** [comment]
> See [#1117 (comment)](https://github.com/Azure/autorest.csharp/pull/1117#discussion_r605981621) for a writeup, but I have architectural concerns that I'd like resolved before landing this.

https://github.com/Azure/autorest.csharp/pull/1117#discussion_r605993594

File: test/AutoRest.TestServer.Tests/TestMgmtParent.cs
Comment (m-nash) [reply test/AutoRest.TestServer.Tests/TestMgmtParent.cs]: ok, you are right this is the one that @allenjzhang talked about which again is something we are looking at separately.  This PR didn't cause it so I don't want to hold up this PR because of it

File: src/AutoRest.CSharp/AutoRest/Communication/StandaloneGeneratorRunner.cs
Comment (ShivangiReja) [inline src/AutoRest.CSharp/AutoRest/Communication/StandaloneGeneratorRunner.cs:158]: Can we add tests for these two properties?

**Comment by chamons** [comment]
I'm not sure that https://github.com/Azure/autorest.csharp/pull/1117#discussion_r605981621 actually has any movement, sounds like "we don't agree this caused/increased it, we want to fix later?".

Putting code in OutputLibraryTestBase to run the entire generator seems very architecturally suspect as well .

File: test/AutoRest.TestServer.Tests/TestMgmtParent.cs
Comment (pakrym) [reply test/AutoRest.TestServer.Tests/TestMgmtParent.cs]: Is there a tracking issue?

File: src/AutoRest.CSharp/Output/Models/Types/ExtensionDetection.cs
Comment (pakrym) [reply src/AutoRest.CSharp/Output/Models/Types/ExtensionDetection.cs:12]: My bad, was under the impression this was an extension method.

File: src/AutoRest.CSharp/Output/Models/Types/ParentDetection.cs
Comment (allenjzhang) [reply src/AutoRest.CSharp/Output/Models/Types/ParentDetection.cs:21]: @pakrym, we are going to have a multitude of RP swaggers to exercise. The current priority is to create a mainline scenario so we can have at ONE swagger to generate usable SDK. This will allow parallel development both on breath coverage of RP swagger and API adjustment.

**Comment by allenjzhang** [comment]
> I'm not sure that [#1117 (comment)](https://github.com/Azure/autorest.csharp/pull/1117#discussion_r605981621) actually has any movement, sounds like "we don't agree this caused/increased it, we want to fix later?".
> 
> Putting code in OutputLibraryTestBase to run the entire generator seems very architecturally suspect as well .

It is now at iteration 16 and we have addressed many of the comments. With few outstanding issues that we have agreed to address with tracking items, it sounds fine particularly given the stage that mgmt track 2 is at now, and with split implementation. We need to be cognizant of work stream when our whole team are working to get a basic implementation out and having to be blocked by single PR. If this is not accepted at this branch level, then we need to re-visit the working branch which would have even bigger implication down the road.

For outputLibraryTestBase, let's discuss in our meeting today.

**Review by chamons — APPROVED** [review]

**Pull request merged**
